### DISCUSSION TITLE
How can I limit search results with threshold?

### DISCUSSION THREAD

**Discussion by psyoungg22** [milvus-io/milvus#31032] [Q&A and General discussion] [answered]
Hello

I am using metric_type as COSINE.
I want to specify threshold to get only limited search results.
What should I do?
Adding radius or range_filter to index_params does not seem to filter.

Did I approach anything wrong?
Below is index_params that I set up

```
            index_params = {
                'metric_type':'COSINE',  # L2, IP, COSINE
                'index_type': "IVF_FLAT",  # IVF_FLAT, IVF_SQ8, IVF_PQ, HNSW, ANNOY
                'params':{"nlist" : 1},
                'radius': threshold, 
                'range_filter' : 1.0 
            }
```

Thank you for your help.

**Comment by xiaofan-luan** [comment]
'radius': threshold, 
                'range_filter' : 1.0 

has nothing to do index, it is search param

**Reply by xiaofan-luan** [reply]
and nlist should not be set to 1.

**Reply by psyoungg22** [reply]
I used like below
Am I missing something?

```
result = collection.search(
      data = [query_emb],  
      anns_field = "Embedding",
      param = index_params)
```


And then, Can you tell me how to set the n_list?I don't understand the n_list well yet.

**Comment by xiaofan-luan** [comment]
for range search, the index is no difference with normal index

**Comment by yhmo** [comment, marked as answer]
**index_params** is for **collection.create_index()**: https://milvus.io/docs/build_index.md
**search_params** is for **collection.search()**: https://milvus.io/docs/search.md
Don't mix up the two interfaces.

For each index, we have index parameters and search parameters: https://milvus.io/docs/index.md
For IVF_FLAT, "nlist" is index parameter, "nprobe" is search paremeter.

Range search: https://milvus.io/docs/within_range.md#Conduct-a-Range-Search
Be careful, with different metric types, the range of "radius" and "range_filter" is different.
For L2 metric type, the range of distance is [0.0, +∞], distance=0.0 means the most similar. You can set `range_filter <= distance < radius`

For IP and COSINE metric type, the range of distance is [0.0, 1.0], distance=1.0 means the most similar. You can set `radius < distance <= range_filter`

Example script:
```
from pymilvus import (
    connections,
    FieldSchema, CollectionSchema, DataType,
    Collection,
    utility,
)

connections.connect(host='localhost', port='19530')
print(utility.get_server_version())

collection_name = "test"
dim = 512
metric_type = "L2"

fields=[
    FieldSchema(name="id", dtype=DataType.INT64, is_primary=True, auto_id=False),
    FieldSchema(name="vector", dtype = DataType.FLOAT_VECTOR, dim=dim),
]

schema = CollectionSchema(fields=fields, enable_dynamic_field=False)

if utility.has_collection(collection_name):
    utility.drop_collection(collection_name)

collection = Collection(name=collection_name, schema=schema)
print(f"Collection '{collection_name}' created")
print(collection.describe())

batch_count = 10000
data = [
    [k for k in range(batch_count)], # id
    [[10*(k+d)/batch_count for d in range(dim)] for k in range(batch_count)], # vector
]
collection.insert(data)
print("insert done")

# index parameters
index_params = {
    'metric_type': metric_type,
    'index_type': "IVF_FLAT",
    'params': {"nlist": 256},
}
collection.create_index(field_name="vector", index_params=index_params)
print("index created")

# load collection into memory
collection = Collection(collection_name)
collection.load()
print("collection loaded")

res = collection.query(expr="", consistency_level="Strong", output_fields=["count(*)"])
print("row count:", res[0]["count(*)"])

# search parameters
params = {
    'metric_type': metric_type,
    "params": {
        # search parameter for IVF_FLAT index
        "nprobe": 10,
        # search for vectors with a distance smaller than a threshold
        "radius": 30,
        # filter out vectors with a distance smaller than or equal to this value
        "range_filter" : 20
    }
}
vector = [0.5 for _ in range(dim)]
results = collection.search(
    data=[vector],
    anns_field="vector",
    param=params,
    limit=500,
)

for hits in results:
    print(f"topk: {len(hits)}")
    for hit in hits:
        print(hit.to_dict())

```

**Reply by psyoungg22** [reply]
Thanks to you, my understanding of milvus seems to have improved. 

It was very helpful. Thank you!
Related links were also very helpful.
